packages/core/src/lib/markdown-it/highlight/HighlightRule.js

```diff
@@ -0,0 +1,82 @@
+const { HighlightRuleComponent } = require('./HighlightRuleComponent.js');
+
+class HighlightRule {
+  constructor(rule) {
+    /**
+     * @type Array<HighlightRuleComponent>
+     */
+    this.rule = rule;
+  }
+
+  isLineNumber() {
+    return this.rule.length === 1 && this.rule[0].isNumber()
+  }
+  
+  isLineSlice() {
+    return this.rule.length === 1 && this.rule[0].isSlice();
+  }
+  
+  isLineRange() {
+    if (this.rule.length !== 2) {
+      return false;
+    }
+
+    const [isFirstNumber, isSecondNumber] = this.rule.map(comp => comp.isNumber());
+    const [isFirstSlice, isSecondSlice] = this.rule.map(comp => comp.isSlice());
+
+    return (isFirstNumber || isFirstSlice) && (isSecondNumber || isSecondSlice);
+  }
+  
+  static parseRule(ruleString) {
+    const components = ruleString.split('-').map(HighlightRuleComponent.parseRuleComponent);
+    return new HighlightRule(components);
+  }
+  
+  offsetLines(offset) {
+    this.rule.map(comp => comp.offsetLines(offset));
+  }
+  
+  shouldApplyHighlight(lineNumber) {
+    const compares = this.rule.map(comp => comp.compareLine(lineNumber));
+    if (this.isLineRange()) {
+      return (compares[0] <= 0) && (compares[1] >= 0);
```

Comment:
could the magic numbers be removed?

Reply:
but if this is a necessary part of the algorithm, maybe you could use some named constant declarations here, which are self-documenting

`const whatIsThis = compares[0]`

Reply:
Yeah, `compareLine` is basically emulating what Java's `compareTo` does, returning either a positive integer, zero, or a negative integer depending on whether the given line number is before, at, or after the component's line number. So I guess it is inevitable to compare whether it's `<= 0` like the above to ensure that the line number is within the bounds.

I'll just name these checks as constant declarations.